File: src/main/java/com/happystudio/testzilla/service/UserService.java
package com.happystudio.testzilla.service;

import java.util.HashMap;
import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.happystudio.testzilla.dao.MailVerificationDao;
import com.happystudio.testzilla.dao.UserDao;
import com.happystudio.testzilla.dao.UserGroupDao;
import com.happystudio.testzilla.model.MailVerification;
import com.happystudio.testzilla.model.User;
import com.happystudio.testzilla.model.UserGroup;
import com.happystudio.testzilla.util.DigestUtils;
import com.happystudio.testzilla.util.MailSender;

/**
 * 用户Service对象. 为Controller提供服务.
 * @author Xie Haozhe
 */
@Service
@Transactional
public class UserService {
	/**
	 * 通过用户名或电子邮件地址获取用户对象.
	 * @param username - 用户名或电子邮件地址
	 * @return 一个User对象或空引用
	 */
	public User getUserUsingUsernameOrEmail(String username) {
		boolean isUsingEmail = username.indexOf('@') != -1;
	    User user = null;
	        
	    if ( !isUsingEmail ) {
	    	user = userDao.getUserUsingUsername(username);
	    } else {
	    	user = userDao.getUserUsingEmail(username);
	    }
	    return user;
	}
	
    /**
     * 验证用户身份是否有效.
     * @param username - 用户名或电子邮件地址
     * @param password - 密码(已使用MD5加密)
     * @return 一个User的对象或空引用
     */
    public HashMap<String, Boolean> isAccountValid(String username, String password) {
        HashMap<String, Boolean> result = new HashMap<String, Boolean>();
        result.put("isUsernameEmpty", username.isEmpty());
        result.put("isPasswordEmpty", password.isEmpty());
        result.put("isAccountValid", false);
        result.put("isSuccessful", false);

        if ( !result.get("isUsernameEmpty") && !result.get("isPasswordEmpty") ) {
        	User user = getUserUsingUsernameOrEmail(username);
        	if ( user != null && user.getPassword().equals(password) ) {
        		result.put("isAccountValid", true);
                result.put("isSuccessful", true);
        	}
        }
        return result;
    }
    
    /**
     * 验证数据并创建新用户.
     * @param username - 用户名
     * @param password - 密码
     * @param confirmPassword - 确认密码
     * @param userGroupSlug - 用户组的唯一英文简称
     * @param realName - 用户真实姓名或公司名称
     * @param email - 电子邮件地址
	 * @param country - 用户所在国家
	 * @param province - 用户所在省份
	 * @param city - 用户所在城市
	 * @param phone - 用户的联系电话
     * @param website - 用户的个人主页
	 * @param isIndividual - 是否为个人用户
     * @param request - HttpRequest对象
     * @return 一个包含若干标志位的JSON数据
     */
    public HashMap<String, Boolean> createUser(String username, String password, 
    		String confirmPassword, String userGroupSlug, String realName, String email, 
    		String country, String province, String city, String phone, String website, 
    		boolean isIndividual) {
    	UserGroup userGroup = getUserGroup(userGroupSlug);
    	boolean isEmailVerified = false;
    	User user = new User(username, DigestUtils.md5Hex(password), userGroup, realName, email, 
    							country, province, city, phone, website, isIndividual, isEmailVerified);
    	HashMap<String, Boolean> result = getJoinResult(user, password, confirmPassword);
    	
    	if ( result.get("isSuccessful") ) {
    		boolean isSuccessful = userDao.createUser(user);
    		result.put("isSuccessful", isSuccessful);
    	}
    	return result;
    }
    
    /**
     * 检查用户所提交的注册信息是否合法.
     * @param user - 待注册的用户对象
     * @param password - 密码(未使用MD5加密)
     * @param confirmPassword - 确认密码
     * @return 一个包含若干标志位的HashMap<String, Boolean>对象
     */
    private HashMap<String, Boolean> getJoinResult(User user, String password, String confirmPassword) {
    	HashMap<String, Boolean> result = new HashMap<String, Boolean>();
    	result.put("isUsernameEmpty", user.getUsername().isEmpty());
    	result.put("isUsernameLegal", isUsernameLegal(user.getUsername()));
    	result.put("isUsernameExists", isUsernameExists(user.getUsername()));
    	result.put("isPasswordEmpty", password.isEmpty());
    	result.put("isPasswordLegal", isPasswordLegal(password));
    	result.put("isPasswordMatched", password.equals(confirmPassword));
    	result.put("isUserGroupLegal", isUserGroupLegal(user.getUserGroup()));
    	result.put("isEmailEmpty", user.getEmail().isEmpty());
    	result.put("isEmailLegal", isEmailLegal(user.getEmail()));
    	result.put("isEmailExists", isEmailExists(user.getEmail()));
    	result.put("isCountryEmpty", user.getCountry().isEmpty());
    	result.put("isCountryLegal", user.getCountry().length() <= 24);
    	result.put("isProvinceEmpty", user.getProvince().isEmpty());
    	result.put("isProvinceLegal", user.getProvince().length() <= 24);
    	result.put("isCityLegal", user.getCity().length() <= 24);
    	result.put("isPhoneEmpty", user.getPhone().isEmpty());
    	result.put("isPhoneLegal", isPhoneLegal(user.getPhone()));
    	result.put("isWebsiteLegal", isWebsiteLegal(user.getWebsite()));
    	
    	boolean isSuccessful = !result.get("isUsernameEmpty")  &&  result.get("isUsernameLegal")   &&
    			               !result.get("isUsernameExists") && !result.get("isPasswordEmpty")   &&
    			                result.get("isPasswordLegal")  &&  result.get("isPasswordMatched") &&
    			                result.get("isUserGroupLegal") && !result.get("isEmailEmpty")      &&
    			                result.get("isEmailLegal")     && !result.get("isEmailExists")     &&
    			               !result.get("isCountryEmpty")   &&  result.get("isCountryLegal")    &&
    			               !result.get("isProvinceEmpty")  &&  result.get("isProvinceLegal")   &&
    			                result.get("isCityLegal")      && !result.get("isPhoneEmpty")      &&
    			                result.get("isPhoneLegal")     &&  result.get("isWebsiteLegal");
    	result.put("isSuccessful", isSuccessful);
    	return result;
    }
    
    /**
     * 检查用户名是否合法.
     * 规则: 用户名应由[A-Za-z0-9_]组成, 以字母起始且长度在6-16个字符.
     * @param username - 用户名
     * @return 用户名是否合法
     */
    private boolean isUsernameLegal(String username) {
    	return username.matches("^[A-Za-z][A-Za-z0-9_]{5,15}$");
    }
    
    /**
     * 检查用户名是否存在.
     * @param username - 用户名
     * @return 用户名是否存在
     */
    private boolean isUsernameExists(String username) {
    	User user = userDao.getUserUsingUsername(username);
    	return user != null;
    }
    
    /**
     * 检查密码是否合法.
     * 规则: 密码的长度在6-16个字符.
     * @param password - 密码(未经MD5加密)
     * @return 密码是否合法
     */
    private boolean isPasswordLegal(String password) {
    	int passwordLength = password.length();
    	return passwordLength >= 6 && passwordLength <= 16;
    }
    
    /**
     * 通过用户组的唯一英文缩写获取用户组对象.
     * @param userGroupSlug - 用户组的唯一英文缩写
     * @return 用户组对象或空引用
     */
    private UserGroup getUserGroup(String userGroupSlug) {
    	UserGroup userGroup = userGroupDao.getUserGroupUsingSlug(userGroupSlug);
    	return userGroup;
    }
    
    /**
     * 检查用户组是否合法.
     * 规则: 用户组对象不是空引用且不是管理员的用户组对象
     * @param userGroup - 用户组的对象
     * @return 用户组是否合法
     */
    private boolean isUserGroupLegal(UserGroup userGroup) {
    	return userGroup != null && !userGroup.getUserGroupSlug().equals("administrator");
    }
    
    /**
     * 检查电子邮件地址是否合法.
     * 规则: 合法的电子邮件地址且长度不超过64个字符.
     * @param email - 电子邮件地址
     * @return 电子邮件地址是否合法
     */
    private boolean isEmailLegal(String email) {
    	int emailLength = email.length();
    	return emailLength <= 64 && email.matches("^[A-Za-z0-9\\._-]+@[A-Za-z0-9_-]+\\.[A-Za-z0-9\\._-]+$");
    }
    
    /**
     * 检查电子邮件地址是否存在.
     * @param email - 电子邮件地址
     * @return 电子邮件地址是否存在
     */
    private boolean isEmailExists(String email) {
    	User user = userDao.getUserUsingEmail(email);
    	return user != null;
    }

    /**
     * 检查联系电话是否合法.
     * 规则: 由数字, +和-组成的字符串且长度不超过24个字符.
     * @param phone - 联系电话
     * @return 联系电话是否合法
     */
    private boolean isPhoneLegal(String phone) {
    	return phone.matches("^(\\+[0-9]+\\-)*[0-9]+$");
    }
    
    /**
     * 检查个人主页的地址是否合法.
     * 规则: 合法的HTTP(S)协议URL且长度不超过64个字符.
     * @param website - 个人主页的地址
     * @return 个人主页的地址是否合法
     */
    private boolean isWebsiteLegal(String website) {
    	int websiteLength = website.length();
    	return website.isEmpty() || 
    		  (websiteLength <= 64 && website.matches("^(http|https):\\/\\/[A-Za-z0-9-]+\\.[A-Za-z0-9_.]+$"));
    }
    
    /**
     * 保存电子邮件随机确认代码.
     * @param email - 用户的电子邮件
     * @param code - 随机生成的确认代码
     */
    public void dumpEmailConfidential(String email, String code) {
    	MailVerification verification = new MailVerification(email, code);
    	mailVerificationDao.createMailVerification(verification);
    }
    
    /**
     * 向新注册的用户(更改电子邮件地址的用户)发送验证邮件.
     * @param username - 用户的用户名
     * @param email - 用户的电子邮件
     * @param code - 随机生成的确认代码
     */
    public void sendActivationMail(String username, String email, String code) {
    	String templatePath = "/verifyEmail.vm";
    	Map<String, Object> model = new HashMap<String, Object>();
    	model.put("username", username);
    	model.put("email", email);
    	model.put("code", code);
    	
    	String subject = "Activate Your Account";
    	String body = mailSender.getMailContent(templatePath, model);
    	mailSender.sendMail(email, subject, body);
    }
    
    /**
     * 自动注入的UserDAO对象.
     */
    @Autowired
    private UserDao userDao;
    
    /**
     * 自动注入的UserGroupDAO对象.
     */
    @Autowired
    private UserGroupDao userGroupDao;
    
    /**
     * 自动注入的MailVerificationDao对象.
     */
    @Autowired
    private MailVerificationDao mailVerificationDao; 
    
    /**
     * 自动注入的MailSender对象.
     * 用于发送电子邮件至用户邮箱.
     */
    @Autowired
    @Qualifier("testzillaMailSender")
    private MailSender mailSender;
}

File: app/service/UserService.php
<?php

use Phalcon\Logger;
use Phalcon\Logger\Adapter\File as FileAdapter;

/**
 * The bussiness layer for accounts.
 *
 * @package TestZilla\Service\UserService
 * @author Xie Haozhe <zjhzxhz@gmail.com>
 */
class UserService extends Service {
    /**
     * Initialize the Service
     */
    public function initialize() {
        parent::initialize();
    }

    /**
     * Get user object using the unique ID of the user.
     * @param  long $uid - the unique ID of the user
     * @return an array contains infomration of the user
     */
    public function getUserUsingUid($uid) {
        $rowSet = $this->getUserObjectUsingUid($uid);

        if ( $rowSet == NULL ) {
            return $NULL;
        }
        return array(
            'uid'       => $rowSet->getUid(),
            'username'  => $rowSet->getUsername(),
            'email'     => $rowSet->getEmail(),
            'userGroup' => array(
                'userGroupId'   => $rowSet->getUserGroup()->getUserGroupId(),
                'userGroupSlug' => $rowSet->getUserGroup()->getUserGroupSlug(),
                'userGroupName' => $rowSet->getUserGroup()->getUserGroupName(),
            ),
        );
    }

    /**
     * Get user object using the unique ID of the user.
     * @param  long $uid - the unique ID of the user
     * @return an expected user object or NULL
     */
    public function getUserObjectUsingUid($uid) {
        return User::findFirst(array(
            'conditions'    => 'uid = ?1',
            'bind'          => array(
                1           => $uid,
            ),
        ));
    }

    /**
     * Get User object using username or email.
     * @param  String $username - the username or email of the user
     * @return an expected user object or NULL
     */
    public function getUserUsingUsernameOrEmail($username) {
        $isUsingEmail = strpos($username, '@');
        $user         = NULL;

        if ( $isUsingEmail === false ) {
            $user     = $this->getUserUsingUsername($username);
        } else {
            $user     = $this->getUserUsingEmail($username);
        }
        return $user;
    }

    /**
     * Get User object using username.
     * @param String $username - the username of the user
     * @return an expected user object or NULL
     */
    public function getUserUsingUsername($username) {
        $user = User::findFirst(array(
            'conditions'    => 'username = ?1',
            'bind'          => array(
                1           => $username,
            ),
        ));
        return $user;
    }

    /**
     * Get User object using email.
     * @param String $email - the email of the user
     * @return an expected user object or NULL
     */
    public function getUserUsingEmail($email) {
        $user = User::findFirst(array(
            'conditions'    => 'email = ?1',
            'bind'          => array(
                1           => $email,
            ),
        ));
        return $user;
    }

    /**
     * Get the meta data of a user.
     * @param  long  $uid - the unique ID of the user
     * @return an array contains all meta data of the user
     */
    public function getUserMetaUsingUid($uid) {
        $userMeta       = array();
        $resultSet      = UserMeta::find(array(
            'conditions'    => 'uid = ?1',
            'bind'          => array(
                1           => $uid,
            ),
        ));

        foreach ( $resultSet as $rowSet ) {
            $key        = $rowSet->getMetaKey();
            $value      = $rowSet->getMetaValue();

            if ( $key == 'socialLinks' ) {
                $value  = (array)json_decode($value);
            }
            $userMeta   = array_merge($userMeta, array(
                $key  => $value
            ));
        }
        return $userMeta;
    }

    /**
     * Verify if the username and password is correct.
     * @param  String  $username - the username of the user
     * @param  String  $password - the password of the user
     * @return an array with data validation result
     */
    public function isAccountValid($username, $password) {
        $result                 = array(
            'isSuccessful'      => false,
            'isUsernameEmpty'   => empty($username),
            'isPasswordEmpty'   => empty($password),
            'isAccountValid'    => false,
        );

        if ( !$result['isUsernameEmpty'] && !$result['isPasswordEmpty'] ) {
            $user = $this->getUserUsingUsernameOrEmail($username);

            if ( $user != null && $user->getPassword() == $password ) {
                $result['isSuccessful']   = true;
                $result['isAccountValid'] = true;
            }
        }
        return $result;
    }

    /**
     * Verify the data and create the account.
     * @param  String $username - the username of the user
     * @param  String $password - the password of the user
     * @param  String $email - the email of the user
     * @param  boolean $isTokenValid - the CSRF token is valid
     * @return an array with data validation result
     */
    public function createAccount($username, $password, $email, $isTokenValid) {
        $result                 = array(
            'isSuccessful'      => false,
            'isUsernameEmpty'   => empty($username),
            'isUsernameLegal'   => $this->isUsernameLegal($username),
            'isUsernameExists'  => $this->isUsernameExists($username),
            'isPasswordEmpty'   => empty($password),
            'isPasswordLegal'   => $this->isPasswordLegal($password),
            'isEmailEmpty'      => empty($email),
            'isEmailLegal'      => $this->isEmailLegal($email),
            'isEmailExists'     => $this->isEmailExists($email),
            'isTokenValid'      => $isTokenValid,
        );
        $result['isSuccessful'] = !$result['isUsernameEmpty']  &&  $result['isUsernameLegal'] &&
                                  !$result['isUsernameExists'] && !$result['isPasswordEmpty'] &&
                                   $result['isPasswordLegal']  && !$result['isEmailEmpty']    &&
                                   $result['isEmailLegal']     && !$result['isEmailExists']   &&
                                   $result['isTokenValid'];
        
        if ( $result['isSuccessful'] ) {
            $defaultUserGroup   = UserGroup::findFirst("user_group_slug = 'user'");
            $user               = new User();
            $user->setUsername($username);
            $user->setPassword(md5($password));
            $user->setEmail($email);
            $user->setUserGroup($defaultUserGroup);
            
            if ( !$user->create() ) {
                $result['isSuccessful']      = false;
            } else {
                $this->updateUserMeta($user, 'registerTime', date('Y-m-d H:i:s'));
                $this->sendVerificationEmail('Welcome to TestZilla', 'HelloMessage', $username, $email);
            }
        }
        return $result;
    }

    /**
     * Change password for user logged in.
     * @param  User   $user            - the user who wants to change password
     * @param  String $oldPassword     - the password the user used now
     * @param  String $newPassword     - the new password
     * @param  String $confirmPassword - the confirmation of the new password
     * @return an array contains data infer whether the password is changed
     */
    public function changePassword($user, $oldPassword, $newPassword, $confirmPassword) {
        $result     = array(
            'isSuccessful'              => false,
            'isUserLoggedIn'            => $user != NULL,
            'isOldPasswordCorrect'      => $user->getPassword() == md5($oldPassword),
            'isNewPasswordEmpty'        => empty($newPassword),
            'isNewPasswordLegal'        => $this->isPasswordLegal($newPassword),
            'isConfirmPasswordMatched'  => $newPassword == $confirmPassword,
        );
        $result['isSuccessful'] =  $result['isUserLoggedIn']     && $result['isOldPasswordCorrect'] &&
                                  !$result['isNewPasswordEmpty'] && $result['isNewPasswordLegal']   &&
                                   $result['isConfirmPasswordMatched'];

        if ( $result['isSuccessful'] ) {
            $user->setPassword(md5($newPassword));
            if ( !$user->update() ) {
                $result['isSuccessful'] = false;
            }
        }
        return $result;
    }

    /**
     * Update profile of user.
     * @param  User   $user        - the user who wants to update profile
     * @param  String $email       - the email of the user
     * @param  String $location    - the location of the user
     * @param  String $website     - the website of the user
     * @param  String $socialLinks - the social links of the user (in JSON format)
     * @param  String $aboutMe     - the introcution of the user
     * @return an array contains information of updateing profile
     */
    public function updateProfile($user, $email, $location, $website, $socialLinks, $aboutMe) {
        $result = array(
            'isSuccessful'      => false,
            'isEmailEmpty'      => empty($email),
            'isEmailLegal'      => $this->isEmailLegal($email),
            'isEmailExists'     => $this->isEmailExistsExceptCurrentUser($user, $email),
            'isLocationLegal'   => mb_strlen($location, 'utf-8') <= 128,
            'isWebsiteLegal'    => $this->isWebsiteLegal($website),
            'isAboutMeLegal'    => mb_strlen($aboutMe, 'utf-8') <= 256,
        );
        $result['isSuccessful'] = !$result['isEmailEmpty']   && $result['isEmailLegal']    &&
                                  !$result['isEmailExists']  && $result['isLocationLegal'] &&
                                   $result['isWebsiteLegal'] && $result['isAboutMeLegal'];

        if ( $result['isSuccessful'] ) {
            $user->setEmail($email);
            $user->update();

            $this->updateUserMeta($user, 'location', $location);
            $this->updateUserMeta($user, 'website', $website);
            $this->updateUserMeta($user, 'socialLinks', $socialLinks);
            $this->updateUserMeta($user, 'aboutMe', $aboutMe);
        }
        return $result;
    }

    /**
     * Update the meta data of the user.
     * @param  User  $user       - the user who wants to update meta data
     * @param  String $metaKey   - the key of the meta
     * @param  String $metaValue - the value of the meta
     */
    private function updateUserMeta($user, $metaKey, $metaValue) {
        $userMeta = UserMeta::findFirst(array(
            'conditions'    => 'uid = ?1 AND meta_key = ?2',
            'bind'          => array(
                1           => $user->getUid(),
                2           => $metaKey,
            ),
        ));

        if ( $userMeta == NULL ) {
            if ( empty($metaValue) ) {
                return;
            }
            $userMeta = new UserMeta();
            $userMeta->setUser($user);
            $userMeta->setMetaKey($metaKey);
            $userMeta->setMetaValue($metaValue);
            $userMeta->create();
        } else {
            $userMeta->setMetaValue($metaValue);
            $userMeta->update();
        }
    }

    /**
     * Verify if the username field is legal.
     * @param  String $username - the username of the user
     * @return if the username field is legal
     */
    private function isUsernameLegal($username) {
        return preg_match('/^[A-Za-z][A-Za-z0-9_]{5,15}$/', $username);
    }

    /**
     * Check if the username has been taken.
     * @param  String $username - the username of the user
     * @return if the username has been taken
     */
    private function isUsernameExists($username) {
        $user = User::findFirst(array(
            'conditions'    => 'username = ?1',
            'bind'          => array(
                1           => $username,
            ),
        ));
        return $user != NULL;
    }

    /**
     * Verify if the password field is legal.
     * @param  String $password - the password of the user
     * @return if the password field is legal
     */
    private function isPasswordLegal($password) {
        $passwordLength = mb_strlen($password, 'utf-8');
        return $passwordLength >= 6 && $passwordLength <= 16;
    }

    /**
     * Verify if the email field is legal.
     * @param  String $email - the email of the user
     * @return if the email field is legal
     */
    private function isEmailLegal($email) {
        return strlen($email) <= 64 && 
               preg_match('/^[A-Za-z0-9\._-]+@[A-Za-z0-9_-]+\.[A-Za-z0-9\._-]+$/', $email);
    }

    /**
     * Check if the email has been taken.
     * @param  String $email - the email of the user
     * @return if the email has been taken
     */
    private function isEmailExists($email) {
        $user = User::findFirst(array(
            'conditions'    => 'email = ?1',
            'bind'          => array(
                1           => $email,
            ),
        ));
        return $user != NULL;
    }

    /**
     * Check if the email has been taken by other users (except current user).
     * @param  User    $user  - current user
     * @param  String  $email - the email of the user wants to use
     * @return whether the email has been taken by other users
     */
    private function isEmailExistsExceptCurrentUser($user, $email) {
        if ( $user->getEmail() == $email ) {
            return false;
        }
        return $this->isEmailExists($email);
    }

    /**
     * Check if the URL of website is legal.
     * @param  String  $website - the URL of the website
     * @return whether the URL of the website is legal
     */
    private function isWebsiteLegal($website) {
        return strlen($website) <= 64 &&
               preg_match('/^(http|https):\/\/[A-Za-z0-9-]+\.[A-Za-z0-9_.]+$/', $website);
    }

    /**
     * Verify if the token and the email address is valid.
     * @param  String  $email - the email address to verify
     * @param  String  $token - the token used for verify the email
     * @return whether the token and the email address is valid
     */
    public function isEmailTokenValid($email, $token) {
        $emailVerification = EmailVerification::findFirst(array(
            'conditions'    => 'email = ?1',
            'bind'          => array(
                1           => $email,
            ),
        ));

        if ( $emailVerification != NULL && 
             $emailVerification->getToken() == $token &&
             strtotime($emailVerification->getExpireTime()) >= strtotime('now') ) {
            return true;
        }
        return false;
    }

    /**
     * Verify the username and email and send verification email.
     * @param  String  $username     - the username of the user
     * @param  String  $email        - the email of the user
     * @param  boolean $isTokenValid - the CSRF token is valid
     * @return an array which contains data infer whether the username and email exists
     */
    public function forgotPassword($username, $email, $isTokenValid) {
        $isUserExists   = false;
        if ( $isTokenValid && !empty($username) && !empty($email) ) {
            $user               = $this->getUserUsingEmail($email);
            if ( $user != NULL && 
                 $user->getUsername() == $username) {
                $isUserExists   = true;
                $this->sendVerificationEmail('Reset Your Password', 'ResetPassword', $username, $email);
            }
        }

        $result         = array(
            'isSuccessful'      => false,
            'isCsrfTokenValid'  => $isTokenValid,
            'isUserExists'      => $isUserExists,
        );
        $result['isSuccessful'] = $result['isCsrfTokenValid'] && $result['isUserExists'];

        return $result;
    }

    /**
     * Send the verification email to the user.
     * @param  String $subject      - the subject of the email
     * @param  String $templateName - the template name of the email
     * @param  String $username     - the username of the user
     * @param  String $email        - the email of the user
     * @return whether the mail is successfully sent
     */
    private function sendVerificationEmail($subject, $templateName, $username, $email) {
        $token              = uniqid();
        $tomorrowDateTime   = date('Y-m-d H:i:s', strtotime('+1 day'));

        $emailVerification  = EmailVerification::findFirst(array(
            'conditions'    => 'email = ?1',
            'bind'          => array(
                1           => $email,
            ),
        ));
        if ( $emailVerification != NULL ) {
            $emailVerification->delete();
        }

        $emailVerification  = new EmailVerification();
        $emailVerification->setEmail($email);
        $emailVerification->setToken($token);
        $emailVerification->setExpireTime($tomorrowDateTime);
        $emailVerification->create();

        $parameters         = array(
            'username'      => $username,
            'email'         => $email,
            'token'         => $token,
        );
        return $this->mailSender->sendMail($email, $subject, $templateName, $parameters);
    }

    /**
     * Reset the password if the email and token is correct.
     * @param  String  $email           - the email address to verify
     * @param  String  $token           - the token used for verify the email
     * @param  String  $newPassword     - the new password
     * @param  String  $confirmPassword - the confirmation of the new password
     * @param  boolean $isTokenValid    - the CSRF token is valid
     * @return an array which contains data infer whether the password is reset
     */
    public function resetPassword($email, $token, $newPassword, $confirmPassword, $isTokenValid) {
        $isEmailTokenValid      = false;
        if ( $isTokenValid ) {
            $emailVerification  = EmailVerification::findFirst(array(
                'conditions'    => 'email = ?1 AND token = ?2 AND expire_time >= ?3',
                'bind'          => array(
                    1           => $email,
                    2           => $token,
                    3           => date('Y-m-d H:i:s', strtotime('now')),
                ),
            ));

            if ( $emailVerification != NULL ) {
                $isEmailTokenValid = true;
            }
        }

        $result                 = array(
            'isSuccessful'              => false,
            'isCsrfTokenValid'          => $isTokenValid,
            'isEmailTokenValid'         => $isEmailTokenValid,
            'isNewPasswordEmpty'        => empty($newPassword),
            'isNewPasswordLegal'        => $this->isPasswordLegal($newPassword),
            'isConfirmPasswordMatched'  => $newPassword == $confirmPassword,
        );
        $result['isSuccessful'] =  $result['isCsrfTokenValid']   && $result['isEmailTokenValid']  && 
                                  !$result['isNewPasswordEmpty'] && $result['isNewPasswordLegal'] && 
                                   $result['isConfirmPasswordMatched'];

        if ( $result['isSuccessful'] ) {
            $user = $this->getUserUsingEmail($email);
            $user->setPassword(md5($newPassword));

            if ( !$user->update() ) {
                $result['isSuccessful'] = false;
            }
        }
        return $result;
    }

    /**
     * The logger of AccountsController.
     * @var Phalcon\Logger\Adapter\File
     */
    private $logger;
}